File: jwj/week08/[14888] 연산자 끼워넣기.py
'''
* Author    : Jang Woo Jin
* Date      : 2024.11.22(Fri)
* Runtime   : 60 ms
* Memory    : 31120 KB 
* Algorithm : Backtracking
'''

import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

N = int(input())
A = list(map(int, input().split()))
operation = list(map(int, input().split()))

MAX = -float("inf") 
MIN = float("inf") 

def recur(idx, result):
    global MAX, MIN
    if idx == N:
        MAX = max(MAX, result)
        MIN = min(MIN, result)
    
    if operation[0] > 0:
        operation[0] -= 1
        recur(idx + 1, result + A[idx])
        operation[0] += 1
    if operation[1] > 0:
        operation[1] -= 1
        recur(idx + 1, result - A[idx])
        operation[1] += 1
    if operation[2] > 0:
        operation[2] -= 1
        recur(idx + 1, result * A[idx])
        operation[2] += 1
    if operation[3] > 0:
        operation[3] -= 1
        recur(idx + 1, int(result / A[idx]))
        operation[3] += 1

recur(1, A[0])
print(MAX)
print(MIN)

File: 이정훈/8주차/[14888] 연산자 끼워넣기.cpp
/*
* Author    : Lee JeongHun
* Date      : 2024.11.21(Thu)
* Runtime   : 0 ms
* Memory    : 2020 KB
* Algorithm : Backtracking
*/

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main()
{
    // 입력 처리
    int n;
    cin >> n;

    vector<int> arr(n);
    for(int i = 0; i < n; ++i)
    {
        cin >> arr[i];
    }

    vector<int> op;
    op.reserve(n - 1);
    for(int i = 0; i < 4; ++i)
    {
        int count;
        cin >> count;

        for(int j = 0; j < count; ++j)
        {
            op.push_back(i);
        }
    }

    int min, max;
    bool init = true;

    do
    {
        int sum = arr[0];
        
        for(int i = 0; i < op.size(); ++i)
        {
            switch(op[i])
            {
                case 0: // +
                sum += arr[i + 1];
                break;

                case 1: // -
                sum -= arr[i + 1];
                break;

                case 2: // *
                sum *= arr[i + 1];
                break;

                case 3: // /
                sum /= arr[i + 1];
                break;
            }
        }

        if(init)
        {
            max = sum;
            min = sum;
            init = false;
        }
        else if(sum < min)
        {
            min = sum;
        }
        else if(sum > max)
        {
            max = sum;
        }

    } while (next_permutation(op.begin(), op.end()));
    
    // 정답 출력
    cout << max << '\n';
    cout << min << '\n';

    return 0;
}